File: local/memory/LibMemory.java
package upc.local.memory;

import jnr.ffi.*;
import jnr.ffi.types.*;

// import upc.local.memory.jnr.Fingerprint;


public class LibMemory {
  public static LibMemoryIface libMemoryInstance = null;
  public static jnr.ffi.Runtime runtime = null;

  public static LibMemoryIface setupLibrary(String[] paths) {
    LibraryLoader<LibMemoryIface> loader = LibraryLoader.create(LibMemoryIface.class);
    for (String path: paths) {
      loader.search(path);
    }
    libMemoryInstance = loader.load("memory");
    runtime = jnr.ffi.Runtime.getRuntime(libMemoryInstance);
    return libMemoryInstance;
  }

  public static class Error extends Exception {
    public Error(String message) {
      super(message);
    }
    public Error(String message, Throwable cause) {
      super(message, cause);
    }
  }

  public static interface LibMemoryIface {
    public ShmKey shm_get_key(ShmGetKeyRequest request);
    public ShmAllocateResult shm_allocate(ShmAllocateRequest request);
    public ShmRetrieveResult shm_retrieve(ShmRetrieveRequest request);
    public ShmDeleteResult shm_delete(ShmDeleteRequest request);
  }

  // common types!
  public static class Fingerprint extends upc.local.memory.jnr.Fingerprint {
    public Fingerprint(byte[] bytes) {
      super(runtime);
      setBytes(bytes);
    }
  }

  public static class ShmKey extends upc.local.memory.jnr.ShmKey {
    public ShmKey(Fingerprint fingerprintArg, long lengthArg) {
      super(runtime);
      setFingerprint(fingerprintArg);
      setLength(lengthArg);
    }

    public ShmKey(ShmKey otherShmKey) {
      super(runtime);
      copyFrom(otherShmKey);
    }
  }

  // shm_get_key() types!
  public static class ShmGetKeyRequest extends Struct {
    public final Pointer source = new Pointer();

    public ShmGetKeyRequest() {
      super(runtime);
    }

    public ShmGetKeyRequest(jnr.ffi.Runtime runtime) {
      super(runtime);
    }

    public ShmGetKeyRequest(jnr.ffi.Pointer sourceArg) throws ShmKeyError {
      super(runtime);
      source.set(sourceArg);
    }

    // @Override
    // public java.lang.String toString() {
    //   return "ShmGetKeyRequest { " + "source = " + source.size() + " }";
    // }
  }

  // shm_allocate() types!
  public static class ShmAllocateRequest extends Struct {
    public final ShmKey key = new ShmKey();
    public final Pointer source = new Pointer();

    public ShmAllocateRequest() {
      super(runtime);
    }

    public ShmAllocateRequest(jnr.ffi.Runtime runtime) {
      super(runtime);
    }

    public ShmAllocateRequest(ShmKey keyArg, jnr.ffi.Pointer sourceArg) throws ShmKeyError {
      super(runtime);
      key.copyFrom(keyArg);
      source.set(sourceArg);
    }
  }

  public static enum ShmAllocateResult_Tag {
    AllocationSucceeded,
    DigestDidNotMatch,
    AllocationFailed,
  }

  public static class AllocationSucceeded_Body extends Struct {
    public final Pointer _0 = new Pointer();

    public AllocationSucceeded_Body() {
      super(runtime);
    }
  }

  public static class DigestDidNotMatch_Body extends Struct {
    public final ShmKey _0 = new ShmKey();

    public DigestDidNotMatch_Body() {
      super(runtime);
    }
  }

  public static class AllocationFailed_Body extends Struct {
    public final Pointer _0 = new Pointer();

    public AllocationFailed_Body() {
      super(runtime);
    }
  }

  public static class ShmAllocateResult_Body extends Union {
    public final AllocationSucceeded_Body allocation_succeeded = new AllocationSucceeded_Body();
    public final DigestDidNotMatch_Body digest_did_not_match = new DigestDidNotMatch_Body();
    public final AllocationFailed_Body allocation_failed = new AllocationFailed_Body();

    public ShmAllocateResult_Body() {
      super(runtime);
    }
  }

  public static class ShmAllocateResult extends Struct {
    public final Enum8<ShmAllocateResult_Tag> tag =
      new Enum8<ShmAllocateResult_Tag>(ShmAllocateResult_Tag.class);
    // Note: this is an anonymous union in test.h, so the `body` identifier does not exist!
    public final ShmAllocateResult_Body body = new ShmAllocateResult_Body();

    public ShmAllocateResult() {
      super(runtime);
    }

    public ShmAllocateResult(jnr.ffi.Runtime runtime) {
      super(runtime);
    }
  }

  // shm_retrieve() types!
  public static class ShmRetrieveRequest extends Struct {
    public final ShmKey key;

    public ShmRetrieveRequest() {
      super(runtime);
      key = new ShmKey();
    }

    public ShmRetrieveRequest(ShmKey keyArg) {
      super(runtime);
      key = keyArg;
    }
  }

  public static enum ShmRetrieveResult_Tag {
    RetrieveSucceeded,
    RetrieveDidNotExist,
    RetrieveInternalError,
  }

  public static class RetrieveSucceeded_Body extends Struct {
    public final Pointer _0 = new Pointer();

    public RetrieveSucceeded_Body() {
      super(runtime);
    }
  }

  public static class RetrieveInternalError_Body extends Struct {
    public final Pointer _0 = new Pointer();

    public RetrieveInternalError_Body() {
      super(runtime);
    }
  }

  public static class ShmRetrieveResult_Body extends Union {
    public final RetrieveSucceeded_Body retrieve_succeeded = new RetrieveSucceeded_Body();
    public final RetrieveInternalError_Body retrieve_internal_error =
      new RetrieveInternalError_Body();

    public ShmRetrieveResult_Body() {
      super(runtime);
    }
  }

  public static class ShmRetrieveResult extends Struct {
    public final Enum8<ShmRetrieveResult_Tag> tag =
      new Enum8<ShmRetrieveResult_Tag>(ShmRetrieveResult_Tag.class);
    // Note: this is an anonymous union in test.h, so the `body` identifier does not exist!
    public final ShmRetrieveResult_Body body = new ShmRetrieveResult_Body();

    public ShmRetrieveResult() {
      super(runtime);
    }

    public ShmRetrieveResult(jnr.ffi.Runtime runtime) {
      super(runtime);
    }
  }

  // shm_delete() types!
  public static class ShmDeleteRequest extends Struct {
    public final ShmKey key;

    public ShmDeleteRequest() {
      super(runtime);
      key = new ShmKey();
    }

    public ShmDeleteRequest(ShmKey keyArg) {
      super(runtime);
      key = keyArg;
    }
  }

  public static enum ShmDeleteResult_Tag {
    DeletionSucceeded,
    DeleteDidNotExist,
    DeleteInternalError,
  }

  public static class DeleteSucceeded_Body extends Struct {
    public final Pointer _0 = new Pointer();

    public DeleteSucceeded_Body() {
      super(runtime);
    }
  }

  public static class DeleteInternalError_Body extends Struct {
    public final Pointer _0 = new Pointer();

    public DeleteInternalError_Body() {
      super(runtime);
    }
  }

  public static class ShmDeleteResult_Body extends Union {
    public final DeleteInternalError_Body delete_internal_error =
      new DeleteInternalError_Body();

    public ShmDeleteResult_Body() {
      super(runtime);
    }
  }

  public static class ShmDeleteResult extends Struct {
    public final Enum8<ShmDeleteResult_Tag> tag =
      new Enum8<ShmDeleteResult_Tag>(ShmDeleteResult_Tag.class);
    // Note: this is an anonymous union in test.h, so the `body` identifier does not exist!
    public final ShmDeleteResult_Body body = new ShmDeleteResult_Body();

    public ShmDeleteResult() {
      super(runtime);
    }

    public ShmDeleteResult(jnr.ffi.Runtime runtime) {
      super(runtime);
    }
  }
}

File: local/memory/LibMemory.scala
package upc.local.memory

import jnr.ffi._


object LibMemory {
  import LibMemoryEnums._

  trait Iface {
    def shm_get_key(request: ShmGetKeyRequest, result: ShmKey): Unit
    def shm_allocate(request: ShmAllocateRequest, result: ShmAllocateResult): Unit
    def shm_retrieve(request: ShmRetrieveRequest, result: ShmRetrieveResult): Unit
    def shm_delete(request: ShmDeleteRequest, result: ShmDeleteResult): Unit
    def shm_free_error_message(error_message: Pointer): Unit
  }

  abstract class FFIError(message: String, cause: Throwable = null)
      extends RuntimeException(message, cause)

  implicit lazy val instance: Iface = {
    val lib_path = "/Users/dmcclanahan/projects/active/upc/local/target/debug"
    val loader = LibraryLoader.create(classOf[Iface])
    lib_path.split(":").foreach(loader.search(_))
    loader.load("memory")
  }
  implicit lazy val runtime = Runtime.getRuntime(instance)

  private[upc] def intoDirectPointer(bytes: Array[Byte]): Pointer = {
    val ptr = Memory.allocateDirect(runtime, bytes.length)
    ptr.put(0, bytes, 0, bytes.length)
    ptr
  }

  // Used in almost every request object.
  class ShmKey(runtime: Runtime = runtime) extends Struct(runtime) {
    import ShmKey._

    val size_bytes = new Unsigned64
    // FIXME: This used to be a separate Fingerprint struct, like in the actual FFI, but it caused
    // errors -- the equivalent of this array kept being zeroed out. Since a struct that has only
    // one element can be replaced precisely with that element in memory, we still have a 1:1
    // mapping.
    val fingerprint = array(new Array[Unsigned8](FINGERPRINT_LENGTH))

    def getSize: Long = size_bytes.get

    lazy val getFingerprintBytes: Array[Byte] = fingerprint.map(u8 => u8.get.toByte)

    def copyKeyFrom(other: ShmKey): Unit = {
      size_bytes.set(other.getSize)
      fingerprint.zip(other.fingerprint).foreach { case (dst, src) =>
        dst.set(src.get)
      }
    }
  }
  object ShmKey {
    val FINGERPRINT_LENGTH: Int = 32

    def apply(sizeBytes: Long, fingerprint: Array[Byte]): ShmKey = {
      val ret = new ShmKey
      ret.size_bytes.set(sizeBytes)
      ret.fingerprint.zip(fingerprint).foreach { case (dst, src) =>
        dst.set(src)
      }
      ret
    }
  }
  case class ShmKeyError(message: String, cause: Throwable = null)
      extends FFIError(message, cause)

  class SharedMemAddressOrErrorMessage(runtime: Runtime = runtime) extends ShmKey(runtime) {
    val address = new Pointer
    val error_message = new Pointer

    // Copy out the memory from the pointer to a directly allocated buffer.
    lazy val getAddressPointer: jnr.ffi.Pointer = {
      val n = getSize.toInt
      val intermediateArray: Array[Byte] = new Array(n)
      address.get.get(0, intermediateArray, 0, n)
      // FIXME: Why do we have to explicitly allocate new direct memory to avoid segfaults? Why
      // can't we just use a pointer directly to that memory (the pointer returned by calling the
      // FFI method)?
      intoDirectPointer(intermediateArray)
    }
  }

  // [Request] for shm_get_key()
  class ShmGetKeyRequest(runtime: Runtime = runtime) extends Struct(runtime) {
    val size = new Unsigned64
    val source = new Pointer
  }
  object ShmGetKeyRequest {
    def apply(sourceArg: Pointer): ShmGetKeyRequest = {
      val ret = new ShmGetKeyRequest
      ret.size.set(sourceArg.size: Long)
      ret.source.set(sourceArg)
      ret
    }
  }

  // [Request] for shm_allocate()
  class ShmAllocateRequest(runtime: Runtime = runtime) extends ShmKey(runtime) {
    val source = new Pointer
  }
  object ShmAllocateRequest {
    def apply(key: ShmKey, source: Pointer): ShmAllocateRequest = {
      val ret = new ShmAllocateRequest
      ret.copyKeyFrom(key)
      ret.source.set(source)
      ret
    }
  }
  // [Result] for shm_allocate()
  class ShmAllocateResult(runtime: Runtime = runtime)
      extends SharedMemAddressOrErrorMessage(runtime) {
    val status: Enum8[ShmAllocateResultStatus_Tag] = new Enum8(classOf[ShmAllocateResultStatus_Tag])
  }

  // [Request] for shm_retrieve()
  class ShmRetrieveRequest(runtime: Runtime = runtime) extends ShmKey(runtime)
  object ShmRetrieveRequest {
    def apply(key: ShmKey): ShmRetrieveRequest = {
      val ret = new ShmRetrieveRequest
      ret.copyKeyFrom(key)
      ret
    }
  }
  // [Result] for shm_retrieve()
  class ShmRetrieveResult(runtime: Runtime = runtime)
      extends SharedMemAddressOrErrorMessage(runtime) {
    val status: Enum8[ShmRetrieveResultStatus_Tag] = new Enum8(classOf[ShmRetrieveResultStatus_Tag])
  }

  // [Request] for shm_delete()
  class ShmDeleteRequest(runtime: Runtime = runtime) extends ShmKey(runtime)
  object ShmDeleteRequest {
    def apply(key: ShmKey): ShmDeleteRequest = {
      val ret = new ShmDeleteRequest
      ret.copyKeyFrom(key)
      ret
    }
  }
  // [Result] for shm_delete()
  class ShmDeleteResult(runtime: Runtime = runtime) extends ShmKey(runtime) {
    val error = new Pointer
    val status: Enum8[ShmDeleteResultStatus_Tag] = new Enum8(classOf[ShmDeleteResultStatus_Tag])
  }
}
